### PR TITLE
[13.0][FIX] product: move product_variant_combination logic from post to pre-migration

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Supersedes https://github.com/OCA/OpenUpgrade/pull/2984.

### CODE REVIEW

File: addons/product/migrations/13.0.1.2/pre-migration.py [hunk 1]
```diff
@@ -192,6 +192,31 @@ def calculate_product_product_combination_indices(env):
     )
 
 
+def create_and_fill_product_variant_combination(env):
+    if not openupgrade.table_exists(env.cr, "product_variant_combination"):
+        openupgrade.logged_query(
+            env.cr, """
+            CREATE TABLE product_variant_combination (
+                product_product_id INTEGER, product_template_attribute_value_id INTEGER)""",
```
Comment (flotho) [inline addons/product/migrations/13.0.1.2/pre-migration.py:200]: If the table exists, will the ORM rebuild the constraints like there https://github.com/OCA/OpenUpgrade/pull/2984/files#diff-b5b5a877946538c32d839ca4c2c087592ad42245ee44aa37fee846755c3f5b68R199 ?

Comment (MiquelRForgeFlow) [reply addons/product/migrations/13.0.1.2/pre-migration.py:200]: Yeah, odoo automatically update constraints

Comment (pedrobaeza) [reply addons/product/migrations/13.0.1.2/pre-migration.py:200]: On v13 that's not true AFAIK.

Comment (MiquelRForgeFlow) [reply addons/product/migrations/13.0.1.2/pre-migration.py:200]: Yes, it's true, see https://github.com/odoo/odoo/blob/13.0/odoo/addons/base/models/ir_model.py#L1310. The reflection is done when odoo is initialized, in https://github.com/odoo/odoo/blob/13.0/odoo/models.py#L2510.

Comment (pedrobaeza) [reply addons/product/migrations/13.0.1.2/pre-migration.py:200]: The methods you are signalling don't say exactly that. Please show me where `_reflect_constraint` is called unconditionally. If I don't remember bad, it was only called when the table is not found.

Comment (MiquelRForgeFlow) [reply addons/product/migrations/13.0.1.2/pre-migration.py:200]: What do you mean? `_reflect_constraint` is called unconditionally in `_reflect_model` in https://github.com/OCA/OpenUpgrade/blob/13.0/odoo/addons/base/models/ir_model.py#L1383, which is called in `_reflect` in https://github.com/odoo/odoo/blob/13.0/odoo/models.py#L382, which is called in `_auto_init`.

Comment (pedrobaeza) [reply addons/product/migrations/13.0.1.2/pre-migration.py:200]: Oh, sorry, I was mixing the FK constraints of the m2m. Indeed that ones are always populated.

**Review by StefanRijnhart — APPROVED** [review]

**Comment by flotho** [comment]
Thanks a lot for this merge !

**Pull request merged**